# Analiza skupa podataka Hotel Booking

U ovom projektu analiziramo skup podataka **Hotel Booking**, koji sadrži informacije o rezervacijama u gradskim i resort hotelima. Cilj rada je da se kroz obradu i analizu podataka utvrdi koji faktori mogu uticati na otkazivanje rezervacije.

Kao ciljna promenljiva koristi se kolona **is_canceled**, koja označava da li je rezervacija otkazana ili nije. Problem koji rešavamo predstavlja binarnu klasifikaciju, jer model treba da predvidi jednu od dve moguće klase: rezervacija je otkazana ili rezervacija nije otkazana.

U toku rada biće urađeni sledeći koraci: učitavanje i pregled podataka, analiza nedostajućih vrednosti, eksplorativna analiza podataka, vizuelizacija odnosa između promenljivih, dodavanje novih promenljivih kroz feature engineering, priprema podataka za modelovanje i testiranje više algoritama mašinskog učenja.

## 1. Učitavanje biblioteka

Na početku učitavamo biblioteke koje su potrebne za rad sa podacima i osnovnu analizu dataset-a.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

pd.set_option("display.max_columns", None)

## 2. Učitavanje skupa podataka

Dataset se učitava iz foldera **data**. Koristi se CSV fajl **hotel_booking.csv**, koji sadrži podatke o hotelskim rezervacijama.

In [2]:
df = pd.read_csv("../data/hotel_booking.csv")

print("Podaci su uspešno učitani.")
print(f"Broj redova: {df.shape[0]}")
print(f"Broj kolona: {df.shape[1]}")

Podaci su uspešno učitani.
Broj redova: 119390
Broj kolona: 36


## 3. Osnovni pregled podataka

U ovom delu proveravamo osnovnu strukturu dataset-a: prvih nekoliko redova, nazive kolona, tipove podataka i osnovne informacije o podacima.

In [3]:
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,country,market_segment,distribution_channel,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,reserved_room_type,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,name,email,phone-number,credit_card
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,0.0,0,BB,PRT,Direct,Direct,0,0,0,C,C,3,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01,Ernest Barnes,Ernest.Barnes31@outlook.com,669-792-1661,************4322
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,0.0,0,BB,PRT,Direct,Direct,0,0,0,C,C,4,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01,Andrea Baker,Andrea_Baker94@aol.com,858-637-6955,************9157
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,0.0,0,BB,GBR,Direct,Direct,0,0,0,A,C,0,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02,Rebecca Parker,Rebecca_Parker@comcast.net,652-885-2745,************3734
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,0.0,0,BB,GBR,Corporate,Corporate,0,0,0,A,A,0,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02,Laura Murray,Laura_M@gmail.com,364-656-8427,************5677
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,0.0,0,BB,GBR,Online TA,TA/TO,0,0,0,A,A,0,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03,Linda Hines,LHines@verizon.com,713-226-5883,************5498


In [4]:
print(df.columns.tolist())

['hotel', 'is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'meal', 'country', 'market_segment', 'distribution_channel', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'reserved_room_type', 'assigned_room_type', 'booking_changes', 'deposit_type', 'agent', 'company', 'days_in_waiting_list', 'customer_type', 'adr', 'required_car_parking_spaces', 'total_of_special_requests', 'reservation_status', 'reservation_status_date', 'name', 'email', 'phone-number', 'credit_card']


In [5]:
df.dtypes

hotel                                 str
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                    str
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                                  str
country                               str
market_segment                        str
distribution_channel                  str
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                    str
assigned_room_type                    str
booking_changes                     int64
deposit_type                          str
agent                             

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

In [7]:
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000
mean,0.370416,104.011416,2016.156554,27.165173,15.798241,0.927599,2.500302,1.856403,0.103890,0.007949,0.031912,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363
std,0.482918,106.863097,0.707476,13.605138,8.780829,0.998613,1.908286,0.579261,0.398561,0.097436,0.175767,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798
min,0.000000,0.000000,2015.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000
25%,0.000000,18.000000,2016.000000,16.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000
50%,0.000000,69.000000,2016.000000,28.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000
75%,1.000000,160.000000,2017.000000,38.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,737.000000,2017.000000,53.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,1.000000,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000


## Kratak zaključak početnog pregleda

Skup podataka sadrži **119390 redova i 36 kolona**, pri čemu svaki red predstavlja jednu hotelsku rezervaciju. Kolone obuhvataju informacije o hotelu, gostima, terminu boravka, načinu rezervisanja, prethodnim rezervacijama i statusu otkazivanja.

Početni pregled pokazuje da skup sadrži i numeričke i kategorijske promenljive. Uočeno je da pojedine numeričke kolone imaju veoma velike maksimalne vrednosti u odnosu na njihove medijane, što može ukazivati na prisustvo netipičnih ili ekstremnih vrednosti. Takođe, broj validnih vrednosti nije isti u svim kolonama, pa je potrebno detaljno proveriti nedostajuće podatke.

U narednim koracima biće analizirane nedostajuće vrednosti, zatim netipične i ekstremne vrednosti, nakon čega će biti urađena detaljnija eksplorativna analiza i ispitivanje odnosa sa ciljnom promenljivom `is_canceled`.

## 4. Analiza nedostajućih vrednosti

Nedostajući podaci mogu biti zapisani kao prazna polja u originalnom CSV fajlu. Međutim, `pandas.read_csv()` ih pri standardnom učitavanju automatski prepoznaje i predstavlja kao `NaN` vrednosti. Zbog toga će se u ovom notebook-u svi nedostajući podaci proveravati pomoću metode `isna()`.

In [14]:
missing_counts = df.isna().sum()
missing_percentages = (missing_counts / len(df) * 100).round(3)

missing_summary = pd.DataFrame({
    "Broj nedostajućih vrednosti": missing_counts,
    "Procenat nedostajućih vrednosti": missing_percentages
})

missing_summary = missing_summary[
    missing_summary["Broj nedostajućih vrednosti"] > 0
].sort_values("Broj nedostajućih vrednosti", ascending=False)

missing_summary

,Broj nedostajućih vrednosti,Procenat nedostajućih vrednosti
company,112593,94.307
agent,16340,13.686
country,488,0.409
children,4,0.003


### Ukupan broj redova sa najmanje jednom nedostajućom vrednošću

Ukupan broj nedostajućih polja predstavlja zbir svih `NaN` vrednosti u dataset-u. Sa druge strane, broj redova sa nedostajućim vrednostima pokazuje koliko rezervacija ima najmanje jedno polje bez podatka.

In [18]:
total_missing_values = df.isna().sum().sum()
rows_with_missing = df.isna().any(axis=1).sum()
rows_with_missing_percentage = (rows_with_missing / len(df) * 100).round(2)

print(f"Ukupan broj nedostajućih polja: {total_missing_values}")
print(f"Broj redova sa najmanje jednom NaN vrednošću: {rows_with_missing}")
print(f"Procenat redova sa najmanje jednom NaN vrednošću: {rows_with_missing_percentage}%")

Ukupan broj nedostajućih polja: 129425
Broj redova sa najmanje jednom NaN vrednošću: 119173
Procenat redova sa najmanje jednom NaN vrednošću: 99.82%


### Zaključak analize nedostajućih vrednosti

Nedostajuće vrednosti pronađene su u kolonama `children`, `country`, `agent` i `company`. Kolona `company` ima ubedljivo najveći procenat nedostajućih vrednosti, dok `agent` ima umeren broj nedostajućih zapisa. Kolone `country` i `children` imaju veoma mali procenat nedostajućih vrednosti.

Visok procenat redova sa najmanje jednom nedostajućom vrednošću najvećim delom je posledica kolone `company`. Zbog toga se ne može zaključiti da je skoro ceo dataset neupotrebljiv, već je potrebno posebno razmotriti značenje nedostajućih vrednosti u svakoj koloni.

U ovom koraku podaci nisu menjani. Način obrade biće određen u kasnijem delu čišćenja, nakon dodatne provere značenja kolona i uticaja nedostajućih vrednosti na dalju analizu.